### PR TITLE
Never convert null to UnsetValue in bindings

### LINKED ISSUE #1798: CheckBox IsThreeState not work as WPF
Create checkbox: 
`     <CheckBox IsChecked="{Binding NeedCompress}" IsThreeState="True"/>`
Set NeedCompress = null, run app. Visual as false:
default
Click on checkbox, NeedCompress set true, visual ok
Click on checkbox, NeedCompress set null, then set false, visual as false (empty)

### LINKED ISSUE #9969: `DefaultValueConverter` assumes that assigning `null` to a value means resetting it to default
**Describe the bug**
`DefaultValueConverter` executes by default on all bindings. When `value` is null and `targetType` is a value type, [it returns `AvaloniaProperty.UnsetValue`](https://github.com/AvaloniaUI/Avalonia/blob/567a38ec7206df80cf81343fcabaa22f8db0e45c/src/Avalonia.Base/Data/Converters/DefaultValueConverter.cs#L31-L34). This value is a special object which indicates that the local value of an `AvaloniaProperty` should be cleared, rather than a new local value being set.

This is bad for two reasons:

1. `Nullable<T>` is a value type, but `null` can be assigned to it.
2. Once a `StyledProperty` has been unset, style or trigger values can become effective. This is unexpected if we have a binding with local priority active.

**To Reproduce**
Add this test to `DefaultValueConverterTests`:

```C#
[Fact]
public void Does_Not_Convert_Null_For_Nullable()
{
    var result = DefaultValueConverter.Instance.Convert(
        null,
        typeof(double?),
        null,
        CultureInfo.InvariantCulture);

    Assert.Equal(null, result);
}
```

The test will fail because `result` is `AvaloniaProperty.UnsetValue`, not null.

**Expected behavior**
The test should pass. Setting a nullable value to null is a valid operation.

`UnsetValue` should be applied only after checking that the value being set is the same as the default value of the `AvaloniaProperty`.

**Desktop (please complete the following information):**

- OS: Windows 10
- Version 11 preview 4

### PR DESCRIPTION
**Pull request by TomEdwardsEnscape**
## What does the pull request do?
This PR prevents unexpected behaviour when a binding produces `null` for a value type property.

## What is the current behavior?
Producing null for a value type property currently causes the value to be unset, even if the property is `Nullable<T>`. This is bad for two reasons:

1. The default value of a `Nullable<T>` property might not be null, in which case it becomes impossible to assign null via a binding! This bug affects `ToggleButton.IsChecked`, and a hack was previously added to fix it in #1799, whereby the default value is null in the property, but `false` if you ask the `getValue` method.
2. Once a `StyledProperty` has been unset, style or trigger values can become effective. This is unexpected if we have a binding with local priority active.

## What is the updated/expected behavior with this PR?
If a binding produces null, null is passed to the target `AvaloniaObject`. It will then do one of two things:

1. If null is valid (reference type or nullable): store null as the new value.
2. If null is not valid: emit an error clear the local value.

Clearing the local value in step 2 is incorrect, since the value could have been generated by a style. WPF instead sets the current value to the property's default (preserving its value source). However, changing this behaviour is out of scope for this PR.

## Checklist

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [x] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Breaking changes
Users who relied on bindings switching null to `UnsetValue` (surely not many?) must now ensure that their bindings explicitly return `UnsetValue`. This can be achieved with a converter.

## Obsoletions / Deprecations
None

## Fixed issues
Fixes #9969
Fixes #1798 (properly)
Prerequisite for #9944

### CODE REVIEW

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0028970-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Review by grokys — CHANGES_REQUESTED** [review]
This PR seems to change the binding behavior to differ from that in WPF, which I think we should really be following for edge cases like this. Or am I misunderstanding?

File: tests/Avalonia.Base.UnitTests/AvaloniaObjectTests_Binding.cs [hunk 1]
```diff
@@ -882,7 +882,7 @@ public void Untyped_LocalValue_Binding_Logs_Invalid_Value_Type()
                 source.OnNext(1.2);
                 source.OnNext("foo");
 
-                Assert.Equal(5.6, target.GetValue(Class1.QuxProperty));
+                Assert.Equal(1.2, target.GetValue(Class1.QuxProperty));
```
[the remark this answers is not in the document]
Comment (tomenscape) [reply tests/Avalonia.Base.UnitTests/AvaloniaObjectTests_Binding.cs:885]: I extended your WPF test a bit. The value changes to the default, BUT it retains its original value source:

```C#
Assert.Equal(BaseValueSource.Local, DependencyPropertyHelper.GetValueSource(target, Class1.DoubleProperty).BaseValueSource);
```

If I provide the initial value via a style binding rather than calling `SetBinding` locally on the object, I get `BaseValueSource.Style` instead.

Looks like this PR just became dependent on #9985!

[the remark this answers is not in the document]
Comment (tomenscape) [reply tests/Avalonia.Base.UnitTests/AvaloniaObjectTests_Binding.cs:885]: It's by no means an implementation detail. It affects the whole outcome of the operation.

When Avalonia/master can't convert the object produced by a binding, it clears the property's local value. This removes any local bindings. But you don't necessarily get the default value, because frames with lower binding priorities can now take effect.

When WPF can't convert an object produced by a binding, it assigns the default value _as_ the current value, with the current priority. This both guarantees that the value is the default, and doesn't remove any bindings.

I think that this branch should start doing what WPF does.

> Oh wait, Avalonia does still have the correct value after these changes

Huh? Now I'm also confused. The current state of the branch shouldn't do anything to the property.

[the remark this answers is not in the document]
Comment (grokys) [reply tests/Avalonia.Base.UnitTests/AvaloniaObjectTests_Binding.cs:885]: Yeah, sorry, I've not been able to free up enough time to review your PRs properly, I will try to find a bit of uninterrupted time to review them properly. Apologies for that.

[the remark this answers is not in the document]
Comment (tomenscape) [reply tests/Avalonia.Base.UnitTests/AvaloniaObjectTests_Binding.cs:885]: No worries. I started implementing the WPF behaviour, but found that the changes would significantly increase the complexity of the PR. So instead I'll revert all my changes in this area. What matters for this PR is that we can assign null to a nullable.

For future reference, there are three classes that implement the setting of prioritised binding values:

* `BindingEntryBase`
* `LocalValueBindingObserver`
* `LocalValueUntypedBindingObserver`

All of these need to be changed at the same time. Currently I have only changed one of them, whoops!

**Comment by grokys** [comment]
@tomenscape could you allow us to push to your branch? Currently I'm unable to update the PR with changes from master.

**Comment by tomenscape** [comment]
The repo is owned by my company, so I shouldn't give external people write permission to our branches on my own initiative. I'll dodge that question for now and update the branch myself.

I'm up for a chat, and before/after that would also be a good time for us to introduce [our company and product](https://enscape3d.com/) in a small group call. The port from WPF we are working on is a large project and forming tighter ties with the Avalonia team has been an item on our agenda for a while. How does that sound to you?

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0029086-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Comment by grokys** [comment]
> I'm up for a chat, and before/after that would also be a good time for us to introduce [our company and product](https://enscape3d.com/) in a small group call. The port from WPF we are working on is a large project and forming tighter ties with the Avalonia team has been an item on our agenda for a while. How does that sound to you?

Sounds good - drop us an email at team@avaloniaui.net!

**Comment by DmitryZhelnin** [comment]
Seem like we can delete [DataGridValueConverter](https://github.com/AvaloniaUI/Avalonia/blob/46bc57943c540339d75dd9ba8546019658548564/src/Avalonia.Controls.DataGrid/DataGridValueConverter.cs) when this PR will be merged. `DataGridValueConverter` is only used to workaround some problems with `Nullable<>` types.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0029409-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Review by grokys — APPROVED** [review]

**Comment by grokys** [comment]
Thanks @tomenscape - LGTM.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0029744-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]
